### ISSUE TITLE
Error on em::lock in transcation

### ISSUE THREAD

**Issue by kgoncharov** [mikro-orm/mikro-orm#182]
**Describe the bug**
Can't set pessimistic lock on write to some entity.

**Stack trace**
```
[2019-10-01 16:26:57.441 +0000] ERROR (21634 on kgoncharov-linux): ExceptionsHandler An open transaction is required for this operation ValidationError: An open transaction is required for this operation
    at Function.transactionRequired (/home/kgoncharov/Projects/PROJECT/node_modules/mikro-orm/dist/utils/ValidationError.js:56:16)
    at UnitOfWork.lockPessimistic (/home/kgoncharov/Projects/PROJECT/node_modules/mikro-orm/dist/unit-of-work/UnitOfWork.js:283:43)
    at UnitOfWork.lock (/home/kgoncharov/Projects/PROJECT/node_modules/mikro-orm/dist/unit-of-work/UnitOfWork.js:98:24)
    at EntityManager.lock (/home/kgoncharov/Projects/PROJECT/node_modules/mikro-orm/dist/EntityManager.js:103:36)
    at MessageService.<anonymous> (/home/kgoncharov/Projects/PROJECT/src/researcher/message/message.service.ts:54:16)
    at Generator.next (<anonymous>)
    at /home/kgoncharov/Projects/PROJECT/src/researcher/message/message.service.ts:20:71
    at new Promise (<anonymous>)
    at __awaiter (/home/kgoncharov/Projects/PROJECT/src/researcher/message/message.service.ts:16:12)
    at /home/kgoncharov/Projects/PROJECT/src/researcher/message/message.service.ts:52:44
    at /home/kgoncharov/Projects/PROJECT/node_modules/mikro-orm/dist/EntityManager.js:97:31
    at /home/kgoncharov/Projects/PROJECT/node_modules/mikro-orm/dist/connections/AbstractSqlConnection.js:29:35
    at /home/kgoncharov/Projects/PROJECT/node_modules/knex/lib/transaction.js:91:24
    at bound (domain.js:415:14)
    at runBound (domain.js:428:12)
    at tryCatcher (/home/kgoncharov/Projects/PROJECT/node_modules/bluebird/js/release/util.js:16:23)
(node:21634) [DEP0097] DeprecationWarning: Using a domain property in MakeCallback is deprecated. Use the async_context variant of MakeCallback or the AsyncResource class instead.
```

**To Reproduce**
Sample code:
```
    const report = await this.em.findOneOrFail(Report, ...);
...
    const message = new Message({report, ...
    });

    await this.em.persistAndFlush(message);
...
    await this.em.transactional(async em => {
      await em.lock(report, LockMode.PESSIMISTIC_WRITE);
      report.messagesCount++;
    });
```

**Expected behavior**
Almost the same code worked on previous version (`2.7.8`)

**Versions**
` "mikro-orm": "^3.0.0-alpha.18",`

**Comment by B4nan** [comment]
Can you try to fetch the entity again inside the transaction? 

    await this.em.transactional(async em => {
      const report1 = await em.findOneOrFail(Report, ...);
      await em.lock(report1, LockMode.PESSIMISTIC_WRITE);
      report1.messagesCount++;
    });

**Comment by B4nan** [comment]
This test is passing in both mysql and postgres, which looks exactly like what you do:

```typescript
  test('lock supports pessimistic locking [pessimistic write]', async () => {
    const author = new Author2('name', 'email');
    await orm.em.persistAndFlush(author);

    const mock = jest.fn();
    const logger = new Logger(mock, true);
    Object.assign(orm.em.config, { logger });

    await orm.em.transactional(async em => {
      await em.lock(author, LockMode.PESSIMISTIC_WRITE);
    });

    expect(mock.mock.calls.length).toBe(3);
    expect(mock.mock.calls[0][0]).toMatch('begin');
    expect(mock.mock.calls[1][0]).toMatch('select 1 from `author2` as `e0` where `e0`.`id` = ? for update');
    expect(mock.mock.calls[2][0]).toMatch('commit');
  });
```

Sounds like an issue with request context to me, will try it in my project as request context issues are pretty hard to cover with tests.

You could do something like this to get around it for now:

```typescript
await this.em.transactional(async em => {
  await (em as any)['unitOfWork'].lock(report, LockMode.PESSIMISTIC_WRITE);
  report.messagesCount++;
});
```

**Comment by kgoncharov** [comment]
> Can you try to fetch the entity again inside the transaction?
> 
> ```
> await this.em.transactional(async em => {
>   const report1 = await em.findOneOrFail(Report, ...);
>   await em.lock(report1, LockMode.PESSIMISTIC_WRITE);
>   report1.messagesCount++;
> });
> ```

doesn't work for me. 
Anyway, I'll try the second workaround shortly
Also, in your example above lock is applied on newly created entity, though in my case `Message` created and persisted, and lock is set on `Report` which already exists and was fetched earlier.

**Comment by B4nan** [comment]
> Also, in your example above lock is applied on newly created entity, though in my case Message created and persisted, and lock is set on Report which already exists and was fetched earlier.

Yeah but the entity is persisted and flushed, and therefore managed by the EM, so it is basically the same case. 

If the workaround will work, then I am pretty sure what's happening.

**Comment by kgoncharov** [comment]
It doesn't, even if I try to combine it with "fetching again" workaround.
```
    await this.em.transactional(async em => {
      const report1 = await this.em.findOneOrFail(
        Report, ...
      );
      await (em as any)['unitOfWork'].lock(report1, LockMode.PESSIMISTIC_WRITE);
      report1.messagesCount++;
    });
```
throws another error on `em::lock` call:
```
ExceptionsHandler Entity Report is not managed. An entity is managed if its fetched from the database or registered as new through EntityManager.persist() ValidationError: Entity Report is not managed. An entity is managed if its fetched from the database or registered as new through EntityManager.persist()
```

**Comment by B4nan** [comment]
That new error is because you are fetching it via the parent EM, instead of the inner one.

**Comment by kgoncharov** [comment]
oh, my fault. let me recheck

**Comment by kgoncharov** [comment]
Same error after changing `this.em` to just `em`.

**Comment by B4nan** [comment]
So it was the problem I had in my mind, should be now fixed in `3.0.0-alpha.19`.

**Comment by kgoncharov** [comment]
It works now. Thanks!
